### PR TITLE
Rewrite of get_gambl_metadata and new helper functions

### PR DESCRIPTION
**Pull request by rdmorin**
The main changes are:

* get_gambl_metadata replaces og_get_gambl_metadata functionality but allows backwards compatibility. This function does a variety of checks for consistency in the metadata/biopsy tables and reports issues it finds. 
* check_gene_expression ensures RNA-seq is prioritized the way we want and always deduplicated
* get_gene_expression now relies on both the new functions and their new features

```
exome_priority_meta = get_gambl_metadata(capture_protocol_priority = "Exome")
2908 capture samples are missing a value for protocol. Assuming Exome.                                                                                 
55 biopsies are missing from the biopsy metadata. This should be fixed!                                                                                
affected cohorts:  DLBCL_Weng,DLBCL_Jain
27 biopsies with discrepancies in the pathology field. This should be fixed!
10 biopsies with discrepancies in the time_point field. This should be fixed!
                                                                                                                                                       
> table(exome_priority_meta$protocol)

      Capture      chromium  ctDNA_Genome         Exome        genome        Genome         polyA         PolyA 
           34             3             5          3213            68          1854            11          1108 
Ribodepletion 
         1235 
```

```
not_exome_priority_meta = get_gambl_metadata(capture_protocol_priority = "Capture")
2908 capture samples are missing a value for protocol. Assuming Exome.                                                                                 
55 biopsies are missing from the biopsy metadata. This should be fixed!                                                                                
affected cohorts:  DLBCL_Weng,DLBCL_Jain
27 biopsies with discrepancies in the pathology field. This should be fixed!
10 biopsies with discrepancies in the time_point field. This should be fixed!
                                                                                                                                                       
> table(not_exome_priority_meta$protocol)

      Capture      chromium  ctDNA_Genome         Exome        genome        Genome         polyA         PolyA 
           52             3             5          3195            68          1854            11          1108 
Ribodepletion 
         1235 
```

### CODE REVIEW

File: R/get_gambl_metadata2.R [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+#' The purpose of this function is to provide the metadata for the non-redundant set of samples from GAMBL, dealing with all 
+#' types of redundancy caused by samples or biopsies that have data from >1 seq_type (genome or capture), different capture protocols (exome or targeted capture) etc.
+#' 
+#' @param dna_seq_type_priority The default is "genome" and the only other option is "capture". For duplicate biopsy_id/patient combinations with different seq_type available, prioritize this seq_type and drop the others. 
+#' @param capture_protocol_priority For duplicate biopsy_id/patient combinations with different seq_type available, prioritize this seq_type and drop the others. 
+#' @param exome_capture_space_priority A vector specifying how to prioritize exome capture space #TODO: implement and test once examples are available
+#' @param mrna_collapse_redundancy Default: TRUE. Set to FALSE to obtain all rows for the mrna seq_type including those that would otherwise be collapsed. 
+#' @param also_normals Set to TRUE to force the return of rows where tissue_status is normal (default is to restrict to tumour)
```
Comment (lkhilton) [inline R/get_gambl_metadata2.R:14]: Can you add a line describing what the `everything` arg does?

File: R/get_gene_expression.R [hunk 2]
```diff
@@ -1,294 +1,211 @@
...
+      expression_wide = select(expression_wide,-gene_id,-hgnc_symbol) %>% 
+        column_to_rownames("ensembl_gene_id") %>%
+        t() %>%
+        as.data.table(keep.rownames=T) %>% 
+        dplyr::rename("sample_id"="rn") %>%
+        left_join(sample_details,.,by="sample_id")
+      return(expression_wide)
+    }else{
+      return(expression_wide)
     }
```
Comment (lkhilton) [inline R/get_gene_expression.R:204]:
```suggestion
    }else{
        if(format == "wide") {
            return(expression_wide)
            } else {
                return(expression_long)
            }
    }
```

Comment (lkhilton) [reply R/get_gene_expression.R:204]: Should we retain the option to return the tidy long format?

**Review by lkhilton — CHANGES_REQUESTED** [review]
One request for missing documentation, and a suggestion to retain the long data format option for get_gene_expression.

File: R/get_gene_expression.R [hunk 3]
```diff
@@ -1,294 +1,211 @@
...
-#' @param from_flatfile Deprecated but left here for backwards compatibility.
-#' @param prioritize_rows_by A named list with one or more vectors. Provide this parameter if you want to filter out 
-#'   duplications (as indicated by the `multi_exp` column of the output table) by prioritizing rows based on the values 
-#'   of columns specified by this parameter. See the **Details** section for more information. 
+#' @param hugo_symbols One or more gene symbols. Cannot be used in conjunction with ensembl_gene_ids. 
+#' @param ensembl_gene_ids One or more ensembl gene IDs. Cannot be used in conjunction with hugo_symbols. 
+#' @param all_genes Set to TRUE to return the full expression data frame without any subsetting (see warnings below). 
+#' @param engine Either readr or grep. The grep engine usually will increase the speed of loading but doesn't work if you want all genes or a very long list.
+#' @param lazy_join If TRUE, your data frame will also have capture_sample_id and genome_sample_id columns provided. See `check_gene_expression` for more information.
```
Comment (lkhilton) [inline R/get_gene_expression.R:21]:
```suggestion
#' @param format Either `wide` or `long`. Wide format returns one column of expression values per gene. Long format returns one column of expression values with the gene stored in a separate column. 
#' @param lazy_join If TRUE, your data frame will also have capture_sample_id and genome_sample_id columns provided. See `check_gene_expression` for more information.
```

File: R/get_gene_expression.R [hunk 4]
```diff
@@ -1,294 +1,211 @@
...
-                               expression_data,
-                               from_flatfile = TRUE,
-                               prioritize_rows_by){
-  
-  # check parameters
-  if(!is.null(join_with)){
-    stopifnot("`join_with` must be one of NULL, \"mrna\", \"genome\", or \"capture\"." = 
-                join_with %in% c("mrna", "genome", "capture"))
+                               verbose=FALSE,
+                               engine="grep",
```
Comment (lkhilton) [inline R/get_gene_expression.R:57]:
```suggestion
                               engine="grep",
                               format="wide",
```

**Review by lkhilton — APPROVED** [review]

**Comment by rdmorin** [comment]
Would someone be able to test this again now that I've patched a few more bugs? It was a bit more involved to implement properly but it looks like the format (long/wide) is working as expected.

**Pull request merged**